sdk/storage/src/common/xml_wrapper.cpp

```diff
@@ -0,0 +1,391 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "common/xml_wrapper.hpp"
+#include <stdexcept>
+#include <iostream>
+
+namespace Azure
+{
+namespace Storage
+{
+namespace Common
+{
+namespace XML
+{
+
+std::string xml_char_to_string(const xmlChar* xml_char)
+{
+  return std::string(reinterpret_cast<const char*>(xml_char));
+}
+
+xml_text_reader_wrapper::xml_text_reader_wrapper(const unsigned char* buffer, unsigned int size)
+{
+  m_reader = xmlReaderForMemory((const char*)buffer, size, NULL, 0, 0);
+}
+
+xml_text_reader_wrapper::~xml_text_reader_wrapper()
+{
+  if (m_reader != nullptr)
+  {
+    xmlFreeTextReader(m_reader);
+    m_reader = nullptr;
+  }
+}
+
+bool xml_text_reader_wrapper::read() { return xmlTextReaderRead(m_reader) == 1; }
+
+unsigned xml_text_reader_wrapper::get_node_type() { return xmlTextReaderNodeType(m_reader); }
+
+bool xml_text_reader_wrapper::is_empty_element()
+{
+  return xmlTextReaderIsEmptyElement(m_reader) == 1;
+}
+
+std::string xml_text_reader_wrapper::get_local_name()
+{
+  auto xml_char = xmlTextReaderLocalName(m_reader);
+  std::string result;
+
+  if (xml_char != nullptr)
+  {
+    result = xml_char_to_string(xml_char);
+    xmlFree(xml_char);
+  }
+
+  return result;
+}
+
+std::string xml_text_reader_wrapper::get_value()
+{
+  auto xml_char = xmlTextReaderValue(m_reader);
+  std::string result;
+
+  if (xml_char != nullptr)
+  {
+    result = xml_char_to_string(xml_char);
+    xmlFree(xml_char);
+  }
+  return result;
+}
+
+bool xml_text_reader_wrapper::move_to_first_attribute()
+{
+  return xmlTextReaderMoveToFirstAttribute(m_reader) == 1;
+}
+
+bool xml_text_reader_wrapper::move_to_next_attribute()
+{
+  return xmlTextReaderMoveToNextAttribute(m_reader) == 1;
+}
+
+xml_element_wrapper::~xml_element_wrapper() {}
+
+xml_element_wrapper::xml_element_wrapper(xmlNode* node)
+{
+  m_ele = node;
+  m_ele->_private = this;
+}
+
+xml_element_wrapper* xml_element_wrapper::add_child(
+    const std::string& name,
+    const std::string& prefix)
+{
+  xmlNs* ns = nullptr;
+  xmlNode* child = nullptr;
+
+  if (m_ele->type != XML_ELEMENT_NODE)
+  {
+    return nullptr;
+  }
+
+  if (!prefix.empty())
+  {
+    ns = xmlSearchNs(m_ele->doc, m_ele, (const xmlChar*)prefix.c_str());
+    if (!ns)
+    {
+      return nullptr;
+    }
+  }
+
+  child = xmlNewNode(ns, (const xmlChar*)name.c_str()); // mem leak?
+
+  if (!child)
+    return nullptr;
+
+  xmlNode* node = xmlAddChild(m_ele, child);
+  if (!node)
+    return nullptr;
+
+  node->_private = new xml_element_wrapper(node);
+  return reinterpret_cast<xml_element_wrapper*>(node->_private);
+}
+
+void xml_element_wrapper::set_namespace_declaration(
+    const std::string& uri,
+    const std::string& prefix)
+{
+  xmlNewNs(
+      m_ele,
+      (const xmlChar*)(uri.empty() ? nullptr : uri.c_str()),
```

Comment:
This block being repeated over and over suggests it should be a function. And the cast has no effect and should be removed.